itest/assets_test.go

```diff
@@ -528,3 +530,90 @@ func testMintAssetsWithTapscriptSibling(t *harnessTest) {
 	t.lndHarness.MineBlocksAndAssertNumTxes(1, 1)
 	t.lndHarness.AssertNumUTXOsWithConf(t.lndHarness.Bob, 1, 1, 1)
 }
+
+// testMintBatchAndTransfer tests that we can mint a batch of assets, observe
+// the finalized batch state, and observe the same batch state after a transfer
+// of an asset from the batch.
+func testMintBatchAndTransfer(t *harnessTest) {
+	ctxb := context.Background()
+	rpcSimpleAssets := MintAssetsConfirmBatch(
+		t.t, t.lndHarness.Miner.Client, t.tapd, simpleAssets,
+	)
+
+	// List the batch right after minting.
+	originalBatches, err := t.tapd.ListBatches(
+		ctxb, &mintrpc.ListBatchRequest{},
+	)
+	require.NoError(t.t, err)
+
+	// We'll make a second node now that'll be the receiver of all the
+	// assets made above.
+	secondTapd := setupTapdHarness(
+		t.t, t, t.lndHarness.Bob, t.universeServer,
+	)
+	defer func() {
+		require.NoError(t.t, secondTapd.stop(!*noDelete))
+	}()
+
+	// In order to force a split, we don't try to send the full first asset.
+	a := rpcSimpleAssets[0]
+	addr, events := NewAddrWithEventStream(
+		t.t, secondTapd, &taprpc.NewAddrRequest{
+			AssetId:      a.AssetGenesis.AssetId,
+			Amt:          a.Amount - 1,
+			AssetVersion: a.Version,
+		},
+	)
+
+	AssertAddrCreated(t.t, secondTapd, a, addr)
+
+	sendResp, sendEvents := sendAssetsToAddr(t, t.tapd, addr)
+	sendRespJSON, err := formatProtoJSON(sendResp)
+	require.NoError(t.t, err)
+
+	t.Logf("Got response from sending assets: %v", sendRespJSON)
+
+	// Make sure that eventually we see a single event for the
+	// address.
+	AssertAddrEvent(t.t, secondTapd, addr, 1, statusDetected)
+
+	// Mine a block to make sure the events are marked as confirmed.
+	MineBlocks(t.t, t.lndHarness.Miner.Client, 1, 1)
+
+	// Eventually the event should be marked as confirmed.
+	AssertAddrEvent(t.t, secondTapd, addr, 1, statusConfirmed)
+
+	// Make sure we have imported and finalized all proofs.
+	AssertNonInteractiveRecvComplete(t.t, secondTapd, 1)
+	AssertSendEventsComplete(t.t, addr.ScriptKey, sendEvents)
+
+	// Make sure the receiver has received all events in order for
+	// the address.
+	AssertReceiveEvents(t.t, addr, events)
+
+	afterBatches, err := t.tapd.ListBatches(
+		ctxb, &mintrpc.ListBatchRequest{},
+	)
+	require.NoError(t.t, err)
+
+	// The batch listed after the transfer should be identical to the batch
+	// listed before the transfer.
+	require.Equal(
+		t.t, len(originalBatches.Batches), len(afterBatches.Batches),
+	)
+
+	originalBatch := originalBatches.Batches[0].Batch
+	afterBatch := afterBatches.Batches[0].Batch
+
+	// Sort the assets from the listed batch before comparison.
+	slices.SortFunc(originalBatch.Assets,
```

Comment:
Why do we sort here? We already do a sort here:
https://github.com/lightninglabs/taproot-assets/blob/e3474192710f3d056230013bf9188d9e182448fd/tapgarden/planter.go#L904-L907
If they aren't similarly sorted, shouldn't we just error out here in the itest?

Reply:
That sort is over batches, but not the assets within. The test assertion was failing due to an unstable ordering of the assets in the batch.

In practice I'm not sure if we need a stable sorting for the assets of the batch, so I left the sorting of returned assets to the itest and not `listBatches()` itself.